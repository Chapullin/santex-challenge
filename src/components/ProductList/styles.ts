import styled from "styled-components";

export const ProductListStyle = styled.div`
    margin-top: 100px;
    padding: 15px 50px;
`;

export const Grid = styled.div`
    display: grid;
    grid-template-columns: repeat(2, 1fr);
    grid-gap: 50px;
`
