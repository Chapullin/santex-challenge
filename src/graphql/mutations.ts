import { gql } from "@apollo/client";

export const mutationAddItemToOrder = gql`
  mutation addItemToOrder($productVariantId: ID!, $quantity: Int!) {
    addItemToOrder(productVariantId: $productVariantId, quantity: $quantity) {
      __typename
      ... on Order {
        subTotal
      }
      __typename
      ... on NegativeQuantityError {
        message
      }
      __typename
      ... on InsufficientStockError {
        message
      }
      __typename
      ... on OrderModificationError {
        message
      }
      __typename
      ... on OrderLimitError {
        message
      }
    }
  }
`;
