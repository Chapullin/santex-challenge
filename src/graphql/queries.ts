import { gql } from "@apollo/client";

export const getProductList = gql`
  query products {
    products {
      items {
        id
        name
        description
      }
    }
  }
`;

