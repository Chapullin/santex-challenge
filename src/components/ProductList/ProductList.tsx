import { useQuery } from "@apollo/client";
import {getProductList} from "../../graphql/queries"

type itemsTypes = {
    id: string | number;
    description: string;
    name: string;
    featuredAsset: assetType
    variants: variantsType[]
};

type assetType = {
    source: string;
};

type variantsType = {
    price: string
}

export function ProductList() {
  const { data, loading, error } = useQuery(getProductList);

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error, something went wrong :(</p>;

  return (<>
      {data.products.items.map(({id, name, description, featuredAsset, variants}: itemsTypes) => (
              <div key={id}>
                  <h3>{name}</h3>
                  <p>AR$ {variants[0].price}</p>
                  <p>{description}</p>
                  <img src={featuredAsset.source} alt=""/>
                  <button>
                      Add item to cart
                  </button>
              </div>
          )
      )}
  </>);
}
