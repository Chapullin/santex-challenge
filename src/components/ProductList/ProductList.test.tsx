import { MockedProvider } from "@apollo/client/testing";
import {act, render, screen, waitFor} from "@testing-library/react";
import { getProductList } from "../../graphql/queries";
import { ProductList } from "./ProductList";

describe("ProductList", () => {
    it("shows an error message when the query fails", async () => {
        const productListMock = {
            request: {
                query: getProductList,
            },
            error: new Error("error"),
        };
        render(
            <MockedProvider mocks={[productListMock]} addTypename={false}>
                <ProductList />
            </MockedProvider>
        );

        await act(() => new Promise((resolve) => setTimeout(resolve, 0)));

        const { getByText } = screen;
        expect(getByText("Error, something went wrong :(")).toBeInTheDocument();
    });
    it("shows the loading message", async () => {
        render(
            <MockedProvider mocks={[]} addTypename={false}>
                <ProductList />
            </MockedProvider>
        );

        const { getByText } = screen;
        expect(getByText("Loading... Please wait")).toBeInTheDocument();
    });
    it('should match the snapshot', async () => {
        const productListMock = {
            request: {
                query: getProductList,
            },
        };
        const {container} = render(
            <MockedProvider mocks={[productListMock]} addTypename={false}>
                <ProductList/>
            </MockedProvider>
        );
        await waitFor(() => {
            expect(container.firstChild).toMatchSnapshot();
        });
    });
});
