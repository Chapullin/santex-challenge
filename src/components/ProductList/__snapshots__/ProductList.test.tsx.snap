// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList should match the snapshot 1`] = `
<div
  class="sc-hLBbgP cMINNu"
>
  Loading... Please wait
</div>
`;
